code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
-            projects.add(teamProject);
-        }
-        
-        projects.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.EXTERNAL_ID.eq(userId)
-                        .and(ACCOUNT.EXTERNAL_ID_TYPE.eq("project:github_user")))
-                        .and(ACCOUNT.STATE.ne("purged"))
-                        .fetch());
-        
+    public List<Account> getAccessibleProjects (Set<ExternalId> externalIds, Boolean isAdmin, Long policyAccount) {
+        List<Account> projects = new ArrayList<>();
+        if(externalIds == null) {
+            return projects;
+        }
+        if (isAdmin) {
+            projects.addAll(create()
+                    .selectFrom(ACCOUNT)
+                    .where(ACCOUNT.KIND.eq(ProjectConstants.TYPE)
+                    .and(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)))
+                    .orderBy(ACCOUNT.ID.asc())
+                    .fetch());
+            return projects;
+        }
+        if (policyAccount != null) {
+            Account account = getAccountById(policyAccount);
+            if (account != null) {
+                if (account.getKind().equalsIgnoreCase(ProjectConstants.TYPE)) {
+                    projects.add(account);
+                }
+            }
+        }
+        //DSL.falseCondition is created so that we can dynamically build a or
+        //Condition without caring what the external Ids are and still make one
+        //Database call.
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(id.getType()))
+                    .and(PROJECT_MEMBER.REMOVED.isNull())
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE)));
+        }
+        SelectQuery query = create().selectQuery();
+        query.addFrom(ACCOUNT);
+        query.addJoin(PROJECT_MEMBER, PROJECT_MEMBER.PROJECT_ID.equal(ACCOUNT.ID));
+        query.addConditions(allMembers);
+        query.setDistinct(true);
+        projects.addAll(query.fetchInto(ACCOUNT));
         return projects;
-        
+    }
+
+    public List<? extends ProjectMember> getProjectMembers(Long id) {
+        return create()
+                .selectFrom(PROJECT_MEMBER)
+                .where(PROJECT_MEMBER.PROJECT_ID.eq(id))
+                .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                .and(PROJECT_MEMBER.REMOVED.isNull())
+                .orderBy(PROJECT_MEMBER.ID.asc()).fetch();
+    }
+
+    @Override
+    public ProjectMember getProjectMember(Long id) {
+        return create()
+                .selectFrom(PROJECT_MEMBER)
+                .where(PROJECT_MEMBER.ID.eq(id)).fetchOne();
+    }
+
+    @Override
+    public Boolean hasAccessToProject (Long projectId, Long policyAccount, Boolean isAdmin, Set<ExternalId> externalIds) {
```

Comment:
James, please change return type to be primitive boolean, as well as projectId, policyAccount. So you don't have to add all bunch of  not null checks here (and null check is missing for policyAccount anyway in isProjectMember)